pkg/awsutils/gen_vpc_ip_limits.go

```diff
@@ -0,0 +1,178 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// +build ignore
+
+// This program generates contributors.go. It can be invoked by running
+// go generate
+package main
+
+import (
+	"log"
+	"os"
+	"sort"
+	"text/template"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+)
+
+const ipLimitFileName = "pkg/awsutils/vpc_ip_resource_limit.go"
+
+type ENILimit struct {
+	InstanceType string
+	ENILimit     int64
+	IPLimit      int64
+}
+
+// Helper function to call the EC2 DescribeInstanceTypes API and generate the IP limit file.
+func main() {
+	// Get session
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+	_, err := sess.Config.Credentials.Get()
+	if err != nil {
+		log.Fatal(err)
+	}
+	svc := ec2.New(sess)
+	describeInstanceTypeInput := &ec2.DescribeInstanceTypesInput{}
+
+	eniLimitMap := make(map[string]ENILimit)
+	fetchMoreInstanceTypes := true
+	for fetchMoreInstanceTypes {
+		output, err := svc.DescribeInstanceTypes(describeInstanceTypeInput)
+		if err != nil {
+			log.Fatal(err)
+		}
+		// We just want the type name, ENI and IP limits
+		for _, info := range output.InstanceTypes {
+			// Ignore any missing values
+			instanceType := aws.StringValue(info.InstanceType)
+			eniLimit := aws.Int64Value(info.NetworkInfo.MaximumNetworkInterfaces)
+			ipLimit := aws.Int64Value(info.NetworkInfo.Ipv4AddressesPerInterface)
+			if instanceType != "" && eniLimit > 0 && ipLimit > 0 {
+				eniLimitMap[instanceType] = newENILimit(instanceType, eniLimit, ipLimit)
+			}
+		}
+		// Paginate to the next request
+		if output.NextToken != nil {
+			describeInstanceTypeInput = &ec2.DescribeInstanceTypesInput{
+				NextToken: output.NextToken,
+			}
+		} else {
+			fetchMoreInstanceTypes = false
+		}
+	}
+
+	// Override faulty values and add missing instance types
+	eniLimitMap = addManualLimits(eniLimitMap)
+
+	// Sort the keys
+	instanceTypes := make([]string, 0)
+	for k := range eniLimitMap {
+		instanceTypes = append(instanceTypes, k)
+	}
+	sort.Strings(instanceTypes)
+	eniLimits := make([]ENILimit, 0)
+	for _, it := range instanceTypes {
+		eniLimits = append(eniLimits, eniLimitMap[it])
+	}
+
+	// Generate the file
+	f, err := os.Create(ipLimitFileName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	limitsTemplate.Execute(f, struct {
+		Timestamp time.Time
+		ENILimits []ENILimit
+	}{
+		Timestamp: time.Now(),
```

Comment:
Might want to do:

```go
Timestamp: time.Now().Format(time.RFC3339),
```

to prevent this in the output file:

```
2019-12-30 11:55:14.0293 -0800 PST m=+0.724514117
```